src/MagicWormhole/Internal/Dispatch.hs

```diff
@@ -0,0 +1,235 @@
+  race (readMessages connState) (action connState)
+  where
+    -- | Read messages from the input channel forever, or until we fail to handle one.
+    readMessages connState = do
+      -- XXX: Maybe one transaction for these?
+      msg <- atomically $ receive connState
+      result <- atomically $ gotMessage connState msg
+      case result of
+        Just err -> do
+          putStrLn @Text $ "[ERROR] " <> show err
+          pure err
+        Nothing -> readMessages connState
+
+rpc :: HasCallStack => ConnectionState -> Messages.ClientMessage -> IO (Either Error Messages.ServerMessage)
+rpc connState req =
+  case expectedResponse req of
+    Nothing ->
+      -- XXX: Pretty sure we can juggle things around at the type level
+      -- to remove this check. (i.e. make a type for RPC requests).
+      pure (Left (ClientError (NotAnRPC req)))
+    Just responseType -> do
+      box' <- atomically $ expectResponse connState responseType
+      case box' of
+        Left clientError -> pure (Left (ClientError clientError))
+        Right box -> do
+          atomically $ send connState req
+          response <- atomically $ waitForResponse connState responseType box
+          pure $ case response of
+                   Messages.Error reason original -> Left (ClientError (BadRequest reason original))
+                   response' -> Right response'
+
+-- | Tell the connection that we expect a response of the given type.
+--
+-- Will fail with a 'ClientError' if we are already expecting a response of this type.
+expectResponse :: ConnectionState -> ResponseType -> STM (Either ClientError (TMVar Messages.ServerMessage))
+expectResponse connState responseType = do
+  pending <- readTVar (pendingVar connState)
+  case HashMap.lookup responseType pending of
+    Nothing -> do
+      box <- newEmptyTMVar
+      writeTVar (pendingVar connState) (HashMap.insert responseType box pending)
+      pure (Right box)
+    Just _ -> pure (Left (AlreadySent responseType))
+
+waitForResponse :: ConnectionState -> ResponseType -> TMVar Messages.ServerMessage -> STM Messages.ServerMessage
+waitForResponse connState responseType box = do
+  response <- takeTMVar box
+  modifyTVar' (pendingVar connState) (HashMap.delete responseType)
+  pure response
+
+-- | Called when we have received a response from the server.
+--
+-- Tells anything waiting for the response that they can stop waiting now.
+gotResponse :: ConnectionState -> ResponseType -> Messages.ServerMessage -> STM (Maybe ServerError)
+gotResponse connState responseType message = do
+  pending <- readTVar (pendingVar connState)
+  case HashMap.lookup responseType pending of
+    Nothing -> pure (Just (ResponseWithoutRequest responseType message))
+    Just box -> do
+      -- TODO: This will block reading from the server (by retrying the
+      -- transaction) if the box is already populated. I don't think we want
+      -- that, but I'm not sure what behavior we do want.
```

Comment:
This *won't* block reading from the server any more. Reading happens in a separate thread that will keep filling the channel with messages while we wait here.